streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java

```diff
@@ -641,4 +669,180 @@ boolean sendingOldValueEnabled() {
         return (ProcessorParameters<K, VR>) kObjectProcessorParameters;
     }
 
+    public <VR, KO, VO> KTable<K, VR> join(final KTable<KO, VO> other,
+                                           final ValueMapper<V, KO> foreignKeyExtractor,
+                                           final ValueJoiner<V, VO, VR> joiner,
+                                           final Materialized<K, VR, KeyValueStore<Bytes, byte[]>> materialized) {
+        Objects.requireNonNull(other, "other can't be null");
+        Objects.requireNonNull(foreignKeyExtractor, "foreignKeyExtractor can't be null");
+        Objects.requireNonNull(joiner, "joiner can't be null");
+        Objects.requireNonNull(materialized, "materialized can't be null");
+
+        return doJoinOnForeignKey(other, foreignKeyExtractor, joiner, new MaterializedInternal<>(materialized));
+    }
+
+
+    private <VR, KO, VO> KTable<K, VR> doJoinOnForeignKey(final KTable<KO, VO> other,
+                                                          final ValueMapper<V, KO> foreignKeyExtractor,
+                                                          final ValueJoiner<V, VO, VR> joiner,
+                                                          final MaterializedInternal<K, VR, KeyValueStore<Bytes, byte[]>> materialized) {
+        ((KTableImpl<?, ?, ?>) other).enableSendingOldValues();
+        enableSendingOldValues();
+
+        final String repartitionProcessorName = builder.newProcessorName(REPARTITION_NAME);
```

Comment:
Just dropping this in before I forget, following KIP-307, we need to provide a mechanism to name these operators and internal states.

Reply:
The wiki says it's currently under discussion - has it already been accepted? Just curious as to if this work needs to be done for 2.3 if it's not going to be a part of it.

Reply:
Woah, it looks like the contributor forgot to update the KIP. It was accepted. I'll update the wiki.

Reply:
Will be adding this in shortly.